pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -20,148 +20,296 @@
+      // a = 1 OR a = 2 OR a = 3 --> can be rewritten as a IN (1, 2, 3)
+      Map.Entry<String, Set<String>> columnAndValues = columnToInEqPredicateValues.entrySet().iterator().next();
+      return buildFilterQueryTreeForColumnAndPredicateInfo(columnAndValues.getKey(), columnAndValues.getValue());
+    }
 
-        if (!rebuildRequired) {
-          // No mutation needed, so just return the same tree
-          return filterQueryTree;
-        } else {
-          // Rebuild the predicates
-          return rebuildFilterPredicate(columnToValues, nonEqualityOperators);
-        }
-      }
-    } else if (filterQueryTree.getChildren() != null) {
-      // Optimize the child nodes, if any
-      applyOptimizationToChildNodes(filterQueryTree);
+    if (!rewriteINEQChildrenOfORRootOperator) {
+      // No predicate rewriting, so just return the same tree
+      return filterQueryTree;
     }
 
-    return filterQueryTree;
-  }
+    // Rewrite the IN/EQUALITY child operators
+    FilterQueryTree optimizedRoot = rebuildFilterPredicate(columnToInEqPredicateValues);
+    // append all non EQ/IN child operators
+    optimizedRoot.getChildren().addAll(nonEqInOperators);
 
-  private void applyOptimizationToChildNodes(FilterQueryTree filterQueryTree) {
-    Iterator<FilterQueryTree> childTreeIterator = filterQueryTree.getChildren().iterator();
-    List<FilterQueryTree> childrenToAdd = null;
+    return optimizedRoot;
+  }
 
-    while (childTreeIterator.hasNext()) {
-      FilterQueryTree childQueryTree = childTreeIterator.next();
-      FilterQueryTree optimizedChildQueryTree = optimize(childQueryTree, filterQueryTree);
+  /**
+   * Optimize FilterQueryTree rooted at AND. We recurse for each
+   * subtree by calling optimize
+   * The root (AND) is kept same and optimize is invoked on
+   * each subtree and updated in the children list of root
+   * @param filterQueryTree root of tree to optimize
+   * @return root of optimized tree
+   *
+   * Examples:
+   *
+   * a IN (1) AND b = 3 -- rewritten to a = 1 AND b = 3
+   * a IN (1) AND b IN (3) -- rewritten to a = 1 AND b = 3
+   * (a = 1 OR a = 2) AND b = 3 -- rewritten to (a IN (1, 2)) AND b = 3
+   * a > 20 AND a < 100 -- not rewritten
+   * a = 100 AND b = 300 -- not rewritten
+   * (a IN (1, 2, 3) OR b IN (4, 5, 6)) AND (a < 20 OR b > 2) -- not rewritten
+   */
+  private FilterQueryTree optimizeTreeRootedAtAND(FilterQueryTree filterQueryTree) {
+    List<FilterQueryTree> children = filterQueryTree.getChildren();
+    for (int i = 0; i < children.size(); i++) {
+      FilterQueryTree childQueryTree = children.get(i);
+      FilterQueryTree optimizedChildQueryTree = optimize(childQueryTree);
       if (childQueryTree != optimizedChildQueryTree) {
-        childTreeIterator.remove();
-        if (childrenToAdd == null) {
-          childrenToAdd = new ArrayList<>();
-        }
-        childrenToAdd.add(optimizedChildQueryTree);
+        children.set(i, optimizedChildQueryTree);
```

Comment:
Directly call `children.set(i, optimize(children.get(i))` to avoid this branching

Reply:
Done